### PR TITLE
Clean up decision_proceduret

### PR DESCRIPTION
**Pull request by peterschrammel**
`decision_proceduret` is only used through `prop_convt`, because it is lacking essential conversion functions. Moving these functions into `decision_proceduret` will enable to replace `prop_convt` by `decision_proceduret` throughout the code base (which is much more intuitive and makes the code more readable).
`prop_convt` will then evolve into interfaces to support advanced features for incremental solving, etc.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
clang-format asks for a minor reformatting.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2f8e740d9be2bb8bbd5f7f73ae674074f108ec8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105913707
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by martin-cs — APPROVED** [review]
Patch does what is described.  I'm not sure why it has been moved out of `util`.  I'm very pro SMT-LIB as an interface to solvers so I am going to be in favour of something conceptually like `prop_convt` becoming our base BUT we should consider if there are other decision procedures we might want to use that don't or can't fit this interface.  For example : I think @dcattaruzza was using mixed integer or linear programming solvers, what about fuzzing approaches ( @cesaro was working on something like this?), are any of the people doing machine leaning things working on this?

**Comment by peterschrammel** [comment]
> there are other decision procedures we might want to use that don't or can't fit this interface. 

There certainly are, but then they also can't be used as drop-in solvers in CBMC as it stands today.
I'm also in favour of building things in a general way, but there must be at least two (current, not future) use cases to enjoy the benefit; otherwise we'll waste our time maintaining generic frameworks that are not used.

File: src/solvers/prop/decision_procedure.h [hunk 1]
```diff
@@ -23,57 +23,61 @@ class tvt;
...
+
 protected:
+  /// Run the decision procedure to solve the problem
   virtual resultt dec_solve() = 0;
 };
 
-inline decision_proceduret &operator<<(
-  decision_proceduret &dest,
-  const exprt &src)
+/// Push Boolean constraint \p src into decision procedure \p dest
```
Comment (smowton) [inline src/solvers/prop/decision_procedure.h]: Push? Is that the same as add?

**Review by smowton — APPROVED** [review]
One query otherwise lgtm

File: src/solvers/prop/decision_procedure.h (hunk 1 above)
Comment (peterschrammel) [reply src/solvers/prop/decision_procedure.h]: Yes. I'll change it to add.

**Comment by martin-cs** [comment]
On Wed, 2019-03-27 at 01:50 -0700, Peter Schrammel wrote:
> There certainly are, but then they also can't be used as drop-in
> solvers in CBMC as it stands today.
> I'm also in favour of building things in a general way, but there
> must be at least two (current, not future) use cases to enjoy the
> benefit; otherwise we'll waste our time maintaining generic
> frameworks that are not used.

Fair.  I guess it's up to them to shout if they want it.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c56f9929b68b4382327e160f9c77153f2720a44a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105996039
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Pull request merged**